core/include/detray/core/detector.hpp

```diff
@@ -273,47 +292,75 @@ class detector {
         return _surfaces.template get<sf_finders::id::e_brute_force>().all();
     }
 
-    /// @returns all surfaces - non-const
+    /// @returns all portals - non-const
+    /// @note Depending on the detector type, this can also container other
+    /// surfaces
     DETRAY_HOST_DEVICE
-    inline auto &surfaces() {
+    inline auto &portals() {
         return _surfaces.template get<sf_finders::id::e_brute_force>().all();
     }
 
-    /// @returns a surface by index - const
+    /// @returns the portals of a given volume @param v - const
+    /// @note Depending on the detector type, this can also container other
+    /// surfaces
+    DETRAY_HOST_DEVICE constexpr auto portals(const volume_type &v) const {
+
+        // Index of the portal collection in the surface store
+        const dindex pt_coll_idx{
+            v.template link<geo_obj_ids::e_portal>().index()};
+
+        const auto &pt_coll =
+            _surfaces
+                .template get<sf_finders::id::e_brute_force>()[pt_coll_idx];
+
+        return pt_coll.all();
+    }
+
+    /// @return the sub-volumes of the detector - const access
     DETRAY_HOST_DEVICE
-    constexpr auto surfaces(geometry::barcode bcd) const
-        -> const surface_type & {
-        return _surfaces.template get<sf_finders::id::e_brute_force>()
-            .all()[bcd.index()];
+    inline auto surface_lookup() const -> const vector_type<surface_type> & {
+        return _surface_lookup;
     }
 
-    /// @returns surfaces of a given type (@tparam sf_id) by volume - const
-    template <geo_obj_ids sf_id = static_cast<geo_obj_ids>(0)>
-    DETRAY_HOST_DEVICE constexpr auto surfaces(
-        const detector_volume<detector> &v) const {
-        return surfaces<sf_id>(v.m_desc);
+    /// @return the sub-volumes of the detector - non-const access
+    DETRAY_HOST_DEVICE
+    inline auto surface_lookup() -> vector_type<surface_type> & {
+        return _surface_lookup;
+    }
+
+    /// @returns a surface using its barcode - const
+    DETRAY_HOST_DEVICE
+    constexpr auto surface(geometry::barcode bcd) const -> surface_type {
+        return _surface_lookup[bcd.index()];
+    }
+
+    /// @returns the overall number of surfaces in the detector
+    DETRAY_HOST_DEVICE
+    constexpr auto n_surfaces() const -> dindex {
+        return static_cast<dindex>(_surface_lookup.size());
     }
 
-    /// @returns surfaces of a given type (@tparam sf_id) by volume - const
-    template <geo_obj_ids sf_id = static_cast<geo_obj_ids>(0)>
-    DETRAY_HOST_DEVICE constexpr auto surfaces(const volume_type &v) const {
-        dindex coll_idx{v.template link<sf_id>().index()};
-        const auto sf_coll =
-            _surfaces.template get<sf_finders::id::e_brute_force>()[coll_idx];
-        return sf_coll.all();
+    /// Add a new surface to the lookup according to its index.
+    DETRAY_HOST
+    constexpr auto add_surface_to_lookup(const surface_type sf) -> void {
```

Comment:
Puts the surface into the lookup at exactly the position that is given by its index. This can come in handy, if the surfaces need to be sorted in a different order than they are read from file/produced by a surface factory.